### DISCUSSION TITLE
restore windows position on multiple workspaces on external monitor

### DISCUSSION THREAD

**Discussion by igalgh** [rxhanson/RectanglePro-Community#151] [Q&A]
I have a question re the "Arrange an entire workspace of apps" which is mentioned in the Pro version:

<img width="762" alt="Screenshot 2022-10-28 at 12 02 13 PM" src="https://user-images.githubusercontent.com/1245766/198548724-de36f21b-8363-4ab4-a037-b5467185a6c0.png">

I work with several workspaces on an external display; each workspace typically contains Webstorm and iTerm2 windows (+possibly google Chrome window)  - essentially one workspace per separate project (typically side by side, either 50-50 or 66-33).

The natural behavior of the OS is when I disconnect/reconnect the external display (or just suspend/wakeup the system) - all windows get "sucked" into the Mac's retina. 

Then, if I want to restore the "One project per workspace" approach I need to spend a good amount of time (~1-2 min) moving all windows back into their "correct" position in their correct project workspace. 

Clearly this is unacceptable penalty for suspend or screen-disconnect event. It's difficult to accept that there are no obvious solutions (2022!).

I'd pay good money for a tool which solves this. 
Is Rectangle Pro such a tool?

**Comment by rxhanson** [comment]
I think you'd want to try it out and see if it does what you're expecting. It's pretty simple to get app layouts set up, see this discussion:
https://github.com/rxhanson/RectanglePro-Community/discussions/111. 

Rectangle Pro has no way to distinguish windows of the same app, other than by their size/location/z-index, so it will just apply the corresponding size/position in z-index order.

If it doesn't do what you're looking for, let me know so I can see if it's possible to adjust it in the way that you're expecting.

**Reply by igalgh** [reply]
Thanks 🙏🏼  
I'll give it a try in the next few days. 
Will check back and lyk

**Reply by igalgh** [reply]
That was fast. 

I signed in after an hour of suspend, the windows locations and position are not as specified in the `Applications` list config.

I have 4 workspaces on the external display ([Samsung CJ791](https://www.samsung.com/us/computing/monitors/curved/34-j791-thunderbolt-3-ultra-wide-screen-curved-monitor-lc34j791wtnxza/)).

All Webstorm and iTerm2 windows moved to the first workspace (the same behavior as without Rectangle).
Basically it seemed the `D+` (or "Display connect") event did not execute.

Is there an option to manually trigger a `D+` event? 

Perhaps I can force the saved `D+`, `D-` event-settings to apply via keyboard shortcut (?)

<img width="427" alt="Screenshot 2022-10-30 at 12 49 28 PM" src="https://user-images.githubusercontent.com/1245766/198874672-1476f33f-2876-449d-9f78-199aa39c5b2a.png">

Exported config file: 
[RectangleProConfig.json.zip](https://github.com/rxhanson/RectanglePro-Community/files/9895644/RectangleProConfig.json.zip)

**Reply by rxhanson** [reply]
This is likely due to some delay with external displays on wake, I’ll put it on my backlog to take a look at this for an upcoming release.

To assign a keyboard shortcut to a group, click on the group row and the popover will appear with a shortcut recorder above the checkboxes.

**Reply by igalgh** [reply]
That would be nice, to have the "Remember windows on which workspace" functionality in one window-management app, which gets applied consistently. 

Personally, I'd rather avoid running multiple window management apps simultaneously, e.g. Rectangle and Display Maid (mentioned earlier).

Talking of which, I followed that advice and run Display Maid for a day: it works, however inconsistent:
In a few resume events the windows ended up at their expected location.
In others, the windows are misplaced: my iTerm/Webstorm windows are moved from the external display to the Retina. 
I can switch workspaces one-by one and hit the "Restore all windows positions" key. However even if doing so, sometimes "Restore" places the iTerm or Webstorm windows in the wrong workspace. Seems that it can't tell the difference between iTerm window (1) and iTerm window (2) etc. It just grabs the first one (in unknown order?) and moves it to the external workspace, although it was in a different workspace before. 

Honestly there are so many features developed nowadays for tools that most people don't use, while the obvious large elephants in the room get ignored. Not casting stones here, it's just a wide phenomena I notice across the IT industry I think needs to improve. 

I'd love to see Rectangle solve it, and would be happy to pay for such functionality

**Reply by marcomusso** [reply]
I'm in the same situation and bought Rectangle Pro after a frustrating period with many other similar solutions. I'm trying D+ events next but for now I agree 100% with @igal1c0de4n. My current solution is indeed a mix of D+ and shortcuts to "force" a specific space layout.
PS: I was also thinking about relative sizes instead of hardcoded w x h... that could adapt better with different external screens... or maybe even rules to scatter windows on x spaces on y monitors so it's possible to have different setups for different... office locations? anyway that's for another thread.

**Reply by rxhanson** [reply]
The current version has another checkbox in the "edit application group" popover for applying the group on wake. I didn't add any time delay into it, so there is still the potential that windows might not get placed properly; it worked well for me but let me know what you find.

@marcomusso You can adjust any x/y coordinates to be relative values by changing them to values less than 1 (decimal or fraction), and you can then specify a display if desired. I have plans to add in a way to toggle between absolute & relative values in a future update.
